File: Projects/JavaProject/src/com/Arrays/ArraySearching.java
package com.Arrays;

import java.util.Arrays;

public class ArraySearching {

    public static void main(String[] args) {
        search1DArray();
        search2DArray();
    }

    // Function to demonstrate searching in 1D array
    public static void search1DArray() {
        System.out.println("------ 1D Array Searching ------");

        int[] arr = {5, 2, 9, 1, 3};
        int key = 9;

        // 1. Linear Search
        boolean found = false;
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] == key) {
                System.out.println("Linear Search: " + key + " found at index " + i);
                found = true;
                break;
            }
        }
        if (!found) {
            System.out.println("Linear Search: " + key + " not found.");
        }

        // 2. Binary Search (Only works on sorted arrays)
        Arrays.sort(arr); // Must sort before binary search
        int index = Arrays.binarySearch(arr, key);
        if (index >= 0) {
            System.out.println("Binary Search: " + key + " found at index " + index + " (after sorting)");
        } else {
            System.out.println("Binary Search: " + key + " not found.");
        }

        System.out.println();
    }

    // Function to demonstrate searching in 2D array
    public static void search2DArray() {
        System.out.println("------ 2D Array Searching ------");

        int[][] matrix = {
            {10, 20, 30},
            {40, 50, 60},
            {70, 80, 90}
        };
        int key = 50;

        // 1. Linear Search in 2D array
        boolean found = false;
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[i].length; j++) {
                if (matrix[i][j] == key) {
                    System.out.println("Found " + key + " at position [" + i + "][" + j + "]");
                    found = true;
                    break;
                }
            }
            if (found) break;
        }

        if (!found) {
            System.out.println(key + " not found in 2D array.");
        }

        System.out.println();
    }
}

File: Projects/PythonProject/Arrays/ArraySearching.py
import bisect

def search_1d_array():
    print("------ 1D Array Searching ------")
    arr = [5, 2, 9, 1, 3]
    key = 9

    # 1. Linear Search
    found = False
    for i in range(len(arr)):
        if arr[i] == key:
            print(f"Linear Search: {key} found at index {i}")
            found = True
            break
    if not found:
        print(f"Linear Search: {key} not found.")

    # 2. Binary Search (requires sorted array)
    arr.sort()
    index = bisect.bisect_left(arr, key)
    if index < len(arr) and arr[index] == key:
        print(f"Binary Search: {key} found at index {index} (after sorting)")
    else:
        print(f"Binary Search: {key} not found.")
    print()

def search_2d_array():
    print("------ 2D Array Searching ------")
    matrix = [
        [10, 20, 30],
        [40, 50, 60],
        [70, 80, 90]
    ]
    key = 50

    found = False
    for i in range(len(matrix)):
        for j in range(len(matrix[i])):
            if matrix[i][j] == key:
                print(f"Found {key} at position [{i}][{j}]")
                found = True
                break
        if found:
            break

    if not found:
        print(f"{key} not found in 2D array.")
    print()

def main():
    search_1d_array()
    search_2d_array()

if __name__ == "__main__":
    main()
